File: 2492-minimum-score-of-a-path-between-two-cities/2492-minimum-score-of-a-path-between-two-cities.cpp
class Solution {
public:
    // dfs 
    void dfs(int u, unordered_map<int, vector<pair<int,int>>> &adj, vector <bool>& visited, int& result){
        visited[u] = true; 

        for(auto & vec : adj[u]){
            int v = vec.first; 
            int dist = vec.second; 
            result = min(result, dist); 
            if(!visited[v]){
                dfs(v,  adj, visited, result); 
            }
        }
    }


    int minScore(int n, vector<vector<int>>& roads) {
        unordered_map <int , vector<pair <int,int>>> adj; 
        for(auto& vec : roads){
            int u = vec[0]; 
            int v = vec[1]; 
            int d = vec[2]; 
            adj[u].push_back({v,d}); 
            adj[v].push_back({u,d}); 
        }

        vector <bool> visited(n, false); 
        int result = INT_MAX; 

        dfs(1, adj, visited, result); 
        return result; 
    }
};

File: 2492-minimum-score-of-a-path-between-two-cities/2492-minimum-score-of-a-path-between-two-cities.py
class Solution:
    def dfs(self, u, adj, visited, result):
        visited[u] = True

        for v, dist in adj[u]:
            result[0] = min(result[0], dist)
            if not visited[v]:
                self.dfs(v, adj, visited, result)

    def minScore(self, n: int, roads: List[List[int]]) -> int:
        adj = defaultdict(list)
        for u, v, d in roads:
            adj[u].append((v, d))
            adj[v].append((u, d))

        visited = [False] * (n + 1)
        result = [sys.maxsize]

        self.dfs(1, adj, visited, result)
        return result[0]